source/SH3/graphics/texture.cpp

```diff
@@ -13,48 +13,329 @@
+
+        if(ret != sizeof(header))
+            die("sh3_texture::Load( ): ReadData( ) != sizeof(header)!");
+    }
+
+    // Now that we're done that, we can check perform some sanity checks on our texture!
+    if(header.texSize != header.texWidth * header.texHeight * (header.bpp / 8))
+    {
+        Log(LogLevel::WARN, "sh3_texture::Load( ): Warning, texSize != width * height * (bpp / 8)!");
+        return; // TODO: Bind a color shader here
+    }
 
-    if(header.bpp == 8)
-        Log(LogLevel::WARN, "WARNING: 8-bit textures are currently unsupported! They WILL look like nothing but pixel garbage.");
+    if(header.bpp == PixelFormat::PALETTE)
+    {
+        palette_info pal_header;
+
+        // First, we need to seek to the palette and read it in.
+        file.Seek(offset + header.batchHeaderSize + header.texFileSize, std::ios_base::beg);
+        file.ReadData(&pal_header, sizeof(pal_header), e);
+
+        // Palette information is stored in blocks (usually of size 64-bytes). We also know how large the
+        // palette is (in bytes, including padding between blocks). From this, we can deduce (with a bit of math)
+        // that the whole palette occupies ~ paletteSize/entrySize bytes of space, contains a total of
+        // entrySize/bypp colors per block, which therefore means we have a total of nBlocks * col_per_block colors,
+        // which equates to about 256-colors in total (which seems accurate for an 8-bit texture).
+
+        std::size_t nBlocks = (pal_header.paletteSize / pal_header.entrySize) / pal_header.bytes_per_pixel;
+        std::size_t colors_per_block = pal_header.entrySize / pal_header.bytes_per_pixel;
+        std::size_t count = 0;
+
+        std::size_t read = 0;
+        palette.resize(colors_per_block * nBlocks * pal_header.bytes_per_pixel);
+
+        //TODO: This doesn't copy palette data correctly!!!
+        while(count < nBlocks)
+        {
+            read = file.ReadData((char*)&palette[16 * count], sizeof(bgra) * colors_per_block, e);
+            file.Seek(192, std::ios_base::cur); // Seek past the padding.
+            count++;
+        }
+
+        // Now that we've completely loaded the palette in its entirety, we can get the 8-bit index value
+        // from the data section of the file and get it's color in the palette!
+
+        //===---THIS IS A CLUSTER FUCK FOR NOW UNTIL WE UNDERSTAND HOW IN THE NAME OF CHRIST THIS WORKS---===//
+        std::vector<std::uint8_t> iBuffer;  // Our index buffer that we put transformed indecies into
+
+        data.resize(header.texWidth * header.texHeight * 3);
+        iBuffer.resize(header.texSize);
+
+        file.Seek(offset + (header.texFileSize - header.texSize), std::ios_base::beg); // Seek to the beginning of data
+
+        std::uint8_t index = 0;
+        rgba pixel;
+        if(header.texWidth > 96) // Apparently this is the distortion flag?!?!
+        {
+            std::uint32_t tempx = 0;
```

Comment:
The compiler will definitively not increase the size of the stack on each iteration for this.
If we were unnecessarily invoking a constructor each iteration, then performance would suffer.